### ISSUE TITLE
iOS PushPage Flickering / Animation Glitching

### ISSUE THREAD

**Issue by Ralpharoo** [OnsenUI/OnsenUI#2324]
__Environment__ 

```
[Core]
    onsenui latest

[Framework]
    angular1 1.5.8

[Framework binding]
    angular-onsenui latest

[Platform]
    Mobile - iOS (seen since ~9/10)

[Browser]
    Mobile - iOS UIWebView / WKWebView
    Mobile - iOS Safari 
```

__Encountered problem__
When pushing a page containing a tabbar element on iPads the right hand edge of the screen flickers black. 

This seems worse on larger screens but can be seen on iPhones with more tabs and images in the toolbar.

This can be resolved by forcing shouldAnimateToolbar to false on all iOS pop calls. 

__How to reproduce__
1. Create the below application (with relevant includes)
2. Open on an iPad within the PhoneGap developer app / Safari
3. Click 'push'
4a. Now; Repeatedly click 'add page 2', 'push', etc.. - no glitch or problems
4b. Now try; Repeatedly click 'Add tab', 'push' - you should see the glitch every-time the page with the tabbar is pushed

Note; this is much much more obvious on slower / older iPads. 

- __HTML__

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
  <meta http-equiv="Content-Security-Policy" content="default-src * data:; style-src * 'unsafe-inline'; script-src * 'unsafe-inline' 'unsafe-eval'">

  <title>OnsenTests</title>

  <script src="./lib/onsenui/js/onsenui.js"></script>
  <script src="./lib/angular/angular-1.5.8.min.js"></script>
  <script src="./lib/onsenui/js/angular-onsenui.js"></script>
  <script src="js/app.js"></script>

  <script>
      var app = angular.module('myApp', ['onsen']);
  </script>

  <link rel="stylesheet" href="./lib/onsenui/css/onsenui.css">
  <link rel="stylesheet" href="./lib/onsenui/css/onsen-css-components.css">

  <style>
    .logo {
      background:
              url(data:image/gif;base64,R0lGODlhPQBEAPeoAJosM//AwO/AwHVYZ/z595kzAP/s7P+goOXMv8+fhw/v739/f+8PD98fH/8mJl+fn/9ZWb8/PzWlwv///6wWGbImAPgTEMImIN9gUFCEm/gDALULDN8PAD6atYdCTX9gUNKlj8wZAKUsAOzZz+UMAOsJAP/Z2ccMDA8PD/95eX5NWvsJCOVNQPtfX/8zM8+QePLl38MGBr8JCP+zs9myn/8GBqwpAP/GxgwJCPny78lzYLgjAJ8vAP9fX/+MjMUcAN8zM/9wcM8ZGcATEL+QePdZWf/29uc/P9cmJu9MTDImIN+/r7+/vz8/P8VNQGNugV8AAF9fX8swMNgTAFlDOICAgPNSUnNWSMQ5MBAQEJE3QPIGAM9AQMqGcG9vb6MhJsEdGM8vLx8fH98AANIWAMuQeL8fABkTEPPQ0OM5OSYdGFl5jo+Pj/+pqcsTE78wMFNGQLYmID4dGPvd3UBAQJmTkP+8vH9QUK+vr8ZWSHpzcJMmILdwcLOGcHRQUHxwcK9PT9DQ0O/v70w5MLypoG8wKOuwsP/g4P/Q0IcwKEswKMl8aJ9fX2xjdOtGRs/Pz+Dg4GImIP8gIH0sKEAwKKmTiKZ8aB/f39Wsl+LFt8dgUE9PT5x5aHBwcP+AgP+WltdgYMyZfyywz78AAAAAAAD///8AAP9mZv///wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAEAAKgALAAAAAA9AEQAAAj/AFEJHEiwoMGDCBMqXMiwocAbBww4nEhxoYkUpzJGrMixogkfGUNqlNixJEIDB0SqHGmyJSojM1bKZOmyop0gM3Oe2liTISKMOoPy7GnwY9CjIYcSRYm0aVKSLmE6nfq05QycVLPuhDrxBlCtYJUqNAq2bNWEBj6ZXRuyxZyDRtqwnXvkhACDV+euTeJm1Ki7A73qNWtFiF+/gA95Gly2CJLDhwEHMOUAAuOpLYDEgBxZ4GRTlC1fDnpkM+fOqD6DDj1aZpITp0dtGCDhr+fVuCu3zlg49ijaokTZTo27uG7Gjn2P+hI8+PDPERoUB318bWbfAJ5sUNFcuGRTYUqV/3ogfXp1rWlMc6awJjiAAd2fm4ogXjz56aypOoIde4OE5u/F9x199dlXnnGiHZWEYbGpsAEA3QXYnHwEFliKAgswgJ8LPeiUXGwedCAKABACCN+EA1pYIIYaFlcDhytd51sGAJbo3onOpajiihlO92KHGaUXGwWjUBChjSPiWJuOO/LYIm4v1tXfE6J4gCSJEZ7YgRYUNrkji9P55sF/ogxw5ZkSqIDaZBV6aSGYq/lGZplndkckZ98xoICbTcIJGQAZcNmdmUc210hs35nCyJ58fgmIKX5RQGOZowxaZwYA+JaoKQwswGijBV4C6SiTUmpphMspJx9unX4KaimjDv9aaXOEBteBqmuuxgEHoLX6Kqx+yXqqBANsgCtit4FWQAEkrNbpq7HSOmtwag5w57GrmlJBASEU18ADjUYb3ADTinIttsgSB1oJFfA63bduimuqKB1keqwUhoCSK374wbujvOSu4QG6UvxBRydcpKsav++Ca6G8A6Pr1x2kVMyHwsVxUALDq/krnrhPSOzXG1lUTIoffqGR7Goi2MAxbv6O2kEG56I7CSlRsEFKFVyovDJoIRTg7sugNRDGqCJzJgcKE0ywc0ELm6KBCCJo8DIPFeCWNGcyqNFE06ToAfV0HBRgxsvLThHn1oddQMrXj5DyAQgjEHSAJMWZwS3HPxT/QMbabI/iBCliMLEJKX2EEkomBAUCxRi42VDADxyTYDVogV+wSChqmKxEKCDAYFDFj4OmwbY7bDGdBhtrnTQYOigeChUmc1K3QTnAUfEgGFgAWt88hKA6aCRIXhxnQ1yg3BCayK44EWdkUQcBByEQChFXfCB776aQsG0BIlQgQgE8qO26X1h8cEUep8ngRBnOy74E9QgRgEAC8SvOfQkh7FDBDmS43PmGoIiKUUEGkMEC/PJHgxw0xH74yx/3XnaYRJgMB8obxQW6kL9QYEJ0FIFgByfIL7/IQAlvQwEpnAC7DtLNJCKUoO/w45c44GwCXiAFB/OXAATQryUxdN4LfFiwgjCNYg+kYMIEFkCKDs6PKAIJouyGWMS1FSKJOMRB/BoIxYJIUXFUxNwoIkEKPAgCBZSQHQ1A2EWDfDEUVLyADj5AChSIQW6gu10bE/JG2VnCZGfo4R4d0sdQoBAHhPjhIB94v/wRoRKQWGRHgrhGSQJxCS+0pCZbEhAAOw==)
              no-repeat
              center center;
      height: 35px;
      margin: 5px;
      background-size: contain;
    }
  </style>

</head>
<body ng-app="myApp">
  <script src="cordova.js"></script>

    <ons-page>
      <ons-splitter class="theme-background">
        <ons-splitter-side
                page="menu.html"
                side="right"
                mode="collapse"
                collapse=""
                width="306px"
                animation="push"
                class="theme-background"
        >

        </ons-splitter-side>
        <ons-splitter-content id="pageContent">
          <ons-navigator var="topnavigator"
                         swipeable
                         swipe-target-width="75px"
                         swipe-threshold="0.3"
                         page="tab.html"
          >



          </ons-navigator>
        </ons-splitter-content>
      </ons-splitter>
    </ons-page>

  <template id="menu.html">
    <ons-page>
      Menu
    </ons-page>
  </template>

  <template id="tab.html">
    <ons-page>
      <ons-toolbar>
        <div class="center">
          <div class="logo"></div>
        </div>
      </ons-toolbar>

      <ons-tabbar swipeable position="auto">
        <ons-tab page="tab1.html" label="Tab 1" icon="ion-home, material:md-home" badge="7" active>
        </ons-tab>
        <ons-tab page="tab2.html" label="Tab 2" icon="md-settings" active-icon="md-face">
        </ons-tab>
        <ons-tab page="tab2.html" label="Tab 3" icon="md-settings" active-icon="md-face">
        </ons-tab>
      </ons-tabbar>
    </ons-page>
  </template>

    <template id="tab1.html">
      <ons-page id="Tab1">
        <p style="text-align: center;">
          This is the first tab.
          <br/><br/>
          <ons-button ng-click="topnavigator.pushPage('page2.html')">Push</ons-button>
        </p>
      </ons-page>
    </template>

  <template id="page2.html">
    <ons-page>

      <ons-toolbar>
        <div class="left">
          <ons-toolbar-button ng-click="topnavigator.popPage()">
            <ons-icon icon="fa-chevron-left"></ons-icon> Back
          </ons-toolbar-button>
        </div>
        <div class="center">
          <div class="logo"></div>
        </div>
      </ons-toolbar>

      <p style="text-align: center;">
        This is the second page.
        <br/><br/>
        <ons-button ng-click="topnavigator.pushPage('page2.html')">Add Page 2</ons-button>
        <br/><br/>
        <ons-button ng-click="topnavigator.pushPage('tab.html')">Add Tab</ons-button>
      </p>
    </ons-page>
  </template>

    <template id="tab2.html">
      <ons-page id="Tab2">
        <p style="text-align: center;">
          This is the second tab.
        </p>
      </ons-page>
    </template>

</body>
</html>
```

**Comment by frandiox** [comment]
@Ralpharoo  e8dd108 actually only fixes this bug if the active tab is the first one (index 0). I could fix it for the other indexes but it was creating another random flickering. I guess this is the most common scenario so let's close this for now :+1:

**Comment by Ralpharoo** [comment]
Hey @frandiox, thanks as always. I am still seeing this one in the latest development build but suspect it is based on the index / hidden tabs.

I have two scenarios here where I can pretty much always see it;

1. I load page with a hidden tabbar (using hide-tabs) while the content is loaded from the database. This always opens on the first tab though. The tabbar tabs could be shown before, or after the push animation finished based on device speed.
2. I often load a tabbar on another index than the first. I can quickly set the active tab once the animation has finished, however, this looks untidy (but happy to do this if only option).

I've checked and this still can be seen occasionally on Chrome but always on UIWebView and WKWebView. What'd you recommend I do?

I can set up another test if you cannot replicate it.

**Comment by Ralpharoo** [comment]
Note; reverting back to 2.9.0 improves things significantly. Also, changing shouldAnimateToolbar on push to false seemingly helps. Not sure if that's related? Cheers

**Comment by frandiox** [comment]
@Ralpharoo 2.9.0 already contains this fix, and 2.9.1 didn't modify anything important for this. I think the only important change in master branch was made in #2316 fix. Do you think that is affecting the flickering?

Perhaps we should provide a way to specify if the toolbar should be animated or not if this improves pushing pages in some situations.

**Comment by Ralpharoo** [comment]
Hi @frandiox - It is still very apparent in 2.9.0, just not as prominent so not sure those changes could make the difference. It seems related to the the device and screen size, on our smaller iPhone X it is hard to spot - but an older iPad really suffers. I've rolled back further to 2.8.3 and its much much better. It only occurs when pushing a tabbar whose active tab is not the first. The other flicker on page push has gone away. Version 2.7.2 is the same but has other issues.

My best guess is that it is a delay between the animation starting and the entering page being shown / created. Before this, it just has a black square. Interestingly, the same occurs on page 'lifts' - expect the space is white. Adding in extra load (i.e. animating the toolbar) then could well be making this slower and therefore worse.

Video of both here (recorded on iPhone 7+ using WKWebView); [291&283.zip](https://github.com/OnsenUI/OnsenUI/files/1729605/291.283.zip)

Pulling the 2.8.3 video apart, you can see that the second load shows a black box before the entering page is made visible. Then the entering page appears and the animation continues as it should. 

Both suffer pretty bad when the first tabbar page is pushed (with active index of 3). The black area almost makes it to about 75% of the screen before content is shown.

If I get time this weekend I'll have a look at this animation / load process but suspect you know it like the back of your hand!

**Comment by frandiox** [comment]
@Ralpharoo I've added a new `static` attribute for the toolbar which will skip navigator animations and thus improve performance. I think the root of this issue is that a tab bar is loading multiple pages at the same time, not only one. Adding a push animation to that might be too much in some situations. Maybe we can try to modify the way pages are loaded and delay hidden pages until animations are finished... That will take some time, though :/

**Comment by Ralpharoo** [comment]
Hi @frandiox - That sounds likely and a good idea. Something to note, the Android slide and push animation is much better, presume it is more lightweight?

I have taken a look at the TabElement _loadPageElement method and implemented something that seemingly works. However, I am not that close to the intricacies of Onsen so not confident it is right!

```
{
    key: '_loadPageElement',
    value: function _loadPageElement(parent, page) {
      var _this3 = this;
      this._hasLoaded = true;
        
      console.log('_loadPageElement::: ' + page);
      return new Promise(function (resolve) {
          var loadPage = function () {
              console.log('_loadPageElement::: loadPage ' + page);
              _this3._pageLoader.load({parent: parent, page: page}, function (pageElement) {
                  console.log('_loadPageElement::: _pageLoader Done');
                  if (!_this3.isActive()) {
                      // Perf, fixes #2324 when active tab is 0
                      pageElement.style.visibility = 'hidden';
                      _this3._tabbar._loadInactive.promise.then(function () {
                          return pageElement.style.visibility = '';
                      });
                  }
                  parent.replaceChild(pageElement, parent.children[_this3.index]); // Ensure position
                  _this3._loadedPage = pageElement;
                  resolve(pageElement);
              });
          }

          //Are we animating the navigator as part of the page creation?
          var onsNavigator = util.findParent(_this3, 'ons-navigator');
          if (onsNavigator && onsNavigator._isRunning && !_this3.isActive()) {
              //Yes - wait for the animation to complete;

              var onsPage = util.findParent(_this3, 'ons-page');

              //Note I had to make this available to reuse previous logic;
              //util._util$3 = util$3;
              var _cleatListenters = util._util$3.onceOnTransitionEnd(onsPage,function () {
                  clearTimeout(_failSafe);
                  loadPage();
              });
              //Failsafe as I am not confident I haven't broken anything else! :)
              var _failSafe = setTimeout(function () {
                  _cleatListenters();
                  loadPage();
              },1500);
          } else {
              //No - No animation or this is the active tab so need to load it ASAP;
              loadPage();
          }

      });
    }
  }
```

From an initial trial, this works well to resolve some of the issue. The console logs fire as expected, the first page is loaded whilst the other two wait for the page to finish it's animation before being loaded.

Is this what you meant?

Cheers,

**Comment by frandiox** [comment]
@Ralpharoo Thanks for giving it a try! Did you see any flickering after that? The concept is similar but I'd like to do it in a different way without getting the navigator inside tab or tabbar. There is a `_show` hook that should be called right after animations so I think we can use that. I'll try to have a look at it this week :+1:

**Comment by frandiox** [comment]
@Ralpharoo Hey, I think something like this commit e89459658e22056fb162c490d2c35cfdd5f27653 should do the lazy load. However, it might not remove the flickering :/

**Comment by karlvr** [comment]
Just leaving a note to say that we spent four of five hours today trying to get to the bottom of this issue. Using Safari on the desktop to debug Safari on the device (iPhone X), we were able to use breakpoints to observe the behaviour. I've added our notes below in case they help anyone. We have not been able to fix this issue, except to reduce any CPU time spent doing anything else while the transition occurs (as above).

The glitch is where the entering page is invisible, and it reveals the leaving page 25% to the left, and a 25% black stripe down the right. The black stripe is the background that `IOSSlideNavigatorAnimator` adds to the DOM. The 25% is the amount the transition in `IOSSlideNavigatorAnimator` moves the leaving page to the left.

Why is the entering page invisible?

The entering page is not removed from the DOM at any time during the transition. We have verified this by setting "node removed" breakpoints on the new page node for the duration of the transition (and after).

Removing the call to `reset()` in `Animit.restoreStyle()` does not have any impact on the glitch. It of course leaves the DOM nodes with the completed transition attributes intact. We thought that perhaps it was the removal of these elements that caused the issue.

We did discover that if some of the elements that are animated in `IOSSlideNavigatorAnimator.push` don't exist, then we still try to animate them, but we then get the timeout called as those animations don't finish. Perhaps Animit could call done immediately if the supplied elements don't have parent nodes. Changing this, however, did not affect the glitch.

Something that did reduce the appearance of glitches was changing `util.onceOnTransitionEnd` in `animit.js` to call the callback after a timeout, e.g. `setTimeout(callback, 0);`. This dramatically reduced but did not rid us of the glitch.

At the end of the day (and it is, the end of the day) it feels like Safari is dropping the ball on rendering the page divs for a few frames while the CPU is busy.

I hope someone else has better luck that we have, or a fix comes to Safari.
